File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = []
        for head in lists:
            while head:  
                heapq.heappush(min_heap, head.val)
                head = head.next
        res = ListNode()
        cur = res
        for i in range(len(min_heap)):
            cur.next = ListNode(heapq.heappop(min_heap))
            cur = cur.next
        return res.next

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.js
// class Node {
//   constructor(val, next = null, random = null) {
//       this.val = val;
//       this.next = next;
//       this.random = random;
//   }
// }

class Solution {
    /**
     * @param {Node} head
     * @return {Node}
     */
    copyRandomList(head) {
        let cur = head
        const map = new Map()
        
        while (cur){
            const copy = new Node(cur.val)
            map.set(cur, copy)

            cur = cur.next
        }
        cur = head
        map.set(null, null)
        while (cur){
            let copy = map.get(cur)
            copy.next = map.get(cur.next)
            copy.random = map.get(cur.random)
            cur = cur.next
        }

        return map.get(head)
    }
}
